receiver/receivercreator/hints.go

```diff
@@ -0,0 +1,178 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package receivercreator // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/receivercreator"
+
+import (
+	"fmt"
+
+	"github.com/go-viper/mapstructure/v2"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v3"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+)
+
+const (
+	// hints prefix
+	otelHints = "io.opentelemetry.discovery"
+
+	// hint suffix for metrics
+	otelMetricsHints = otelHints + ".metrics"
+
+	// hints definitions
+	discoveryEnabledHint = "enabled"
+	scraperHint          = "scraper"
+	configHint           = "config"
+)
+
+// hintsTemplatesBuilder creates configuration templates from provided hints.
+type hintsTemplatesBuilder interface {
+	createReceiverTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error)
+}
+
+// k8sHintsBuilder creates configurations from hints provided as Pod's annotations.
+type k8sHintsBuilder struct {
+	logger          *zap.Logger
+	config          DiscoveryConfig
+	ignoreReceivers map[string]bool
+}
+
+var _ hintsTemplatesBuilder = (*k8sHintsBuilder)(nil)
+
+func createK8sHintsBuilder(config DiscoveryConfig, logger *zap.Logger) k8sHintsBuilder {
+	ignoreReceivers := make(map[string]bool, len(config.IgnoreReceivers))
+	for _, r := range config.IgnoreReceivers {
+		ignoreReceivers[r] = true
+	}
+	return k8sHintsBuilder{
+		logger:          logger,
+		config:          config,
+		ignoreReceivers: ignoreReceivers,
+	}
+}
+
+// createReceiverTemplateFromHints creates a receiver configuration based on the provided hints.
+// Hints are extracted from Pod's annotations.
+// Scraper configurations are only created for Port Endpoints.
+// TODO: Log receiver configurations are only created for Pod Container Endpoints.
+func (builder *k8sHintsBuilder) createReceiverTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error) {
+	var endpointType string
+	var pod observer.Pod
+
+	endpointType = getStringEnv(env, "type")
+	if endpointType == "" {
+		return nil, fmt.Errorf("could not get endpoint type: %v", zap.Any("env", env))
+	}
+
+	if endpointType != string(observer.PortType) {
+		return nil, nil
+	}
+
+	builder.logger.Debug("handling hints for added endpoint", zap.Any("env", env))
+
+	if endpointPod, ok := env["pod"]; ok {
+		err := mapstructure.Decode(endpointPod, &pod)
+		if err != nil {
+			return nil, fmt.Errorf("could not extract endpoint's pod: %v", zap.Any("endpointPod", pod))
+		}
+		if pod.Annotations != nil {
+			if len(pod.Annotations) == 0 {
+				return nil, nil
+			}
+		}
```

Comment:
```suggestion
		if len(pod.Annotations) == 0 {
			return nil, nil
		}
```
"Slice, or map: the number of elements in v; if v is nil, len(v) is zero." reference - https://pkg.go.dev/builtin#len